src/util/buf.rs

```diff
@@ -0,0 +1,560 @@
+    eq: Some(grpc_slice_default_eq_impl),
+    hash: Some(grpc_slice_default_hash_impl),
+};
+
+impl From<Vec<u8>> for GrpcSlice {
+    /// Converts a `Vec<u8>` into `GrpcSlice`.
+    ///
+    /// There will be a small allocation for custom vtable if the data length
+    /// is too large (23 bytes currently).
+    #[inline]
+    fn from(mut v: Vec<u8>) -> GrpcSlice {
+        let mut slice = GrpcSlice::default();
+        if v.is_empty() {
+            return slice;
+        } else if v.len() <= mem::size_of_val(unsafe { &slice.0.data.inlined.bytes }) {
+            unsafe {
+                slice.0.data.inlined.length = v.len() as u8;
+                slice.0.data.inlined.bytes[..v.len()].copy_from_slice(&v);
+            }
+            return slice;
+        }
+        unsafe {
+            slice.0.data = grpc_slice_grpc_slice_data {
+                refcounted: grpc_slice_grpc_slice_data_grpc_slice_refcounted {
+                    bytes: v.as_mut_ptr(),
+                    length: v.len(),
+                },
+            };
+
+            let mut ref_count = Box::new(VecRefCount {
+                rc: grpc_slice_refcount {
+                    vtable: &VEC_REF_COUNT_VTABLE,
+                    sub_refcount: ptr::null_mut(),
+                },
+                refs: AtomicUsize::new(1),
+                v,
+            });
+            ref_count.rc.sub_refcount = &mut ref_count.rc;
+            slice.0.refcount = Box::into_raw(mem::transmute(ref_count));
+            slice
+        }
+    }
+}
+
+/// A collection of `GrpcBytes`.
+#[repr(C)]
+pub struct GrpcByteBuffer(*mut grpc_byte_buffer);
+
+impl GrpcByteBuffer {
+    #[inline]
+    pub unsafe fn from_raw(ptr: *mut grpc_byte_buffer) -> GrpcByteBuffer {
+        GrpcByteBuffer(ptr)
+    }
+}
+
+impl<'a> From<&'a [GrpcSlice]> for GrpcByteBuffer {
+    /// Create a buffer from the given slice array.
+    ///
+    /// A buffer is allocated for the whole slice array, and every slice will
+    /// be `Clone::clone` into the buffer.
+    fn from(slice: &'a [GrpcSlice]) -> Self {
+        unsafe {
+            let s: &[grpc_slice] = &*(slice as *const _ as *const [grpc_slice]);
+            // hack: see From<&GrpcSlice>.
+            GrpcByteBuffer(grpc_raw_byte_buffer_create(s.as_ptr() as _, s.len()))
+        }
+    }
+}
+
+impl<'a> From<&'a GrpcSlice> for GrpcByteBuffer {
+    /// Create a buffer from the given single slice.
+    ///
+    /// A buffer, which length is 1, is allocated for the slice.
+    #[allow(clippy::cast_ref_to_mut)]
+    fn from(s: &'a GrpcSlice) -> GrpcByteBuffer {
+        unsafe {
+            // hack: buffer_create accepts an mutable pointer to indicate it mutate
+            // ref count. Ref count is recorded by atomic variable, which is considered
+            // `Sync` in rust. This is an interesting difference in what is *mutable*
+            // between C++ and rust.
+            let s = &mut *(s as *const GrpcSlice as *mut GrpcSlice);
```

Comment:
This is UB; you should an `UnsafeCell` here in order to get the `*mut` pointer

Reply:
`UnsafeCell` will make the implement very verbose. This is the only place that does the hack, and it's put inside unsafe already.

Reply:
It does unfortunately make it more verbose but I think that casting like this is UB, whether it is safe or not because it can break LLVM's alias analysis. Adding the `UnsafeCell` means rustc will not tell LLVM that `s` is unique (I'm not 100% sure, I know this is true for `&` pointers, but I can't find documentation about `*` pointers, however I think the compiler treats them the same and it is better to be safe, I think).